File: 2023/UICPC Regional/Round #8/solutions/Outsourcing/solutions/tobi.java
// @EXPECTED_RESULTS@: CORRECT

import java.io.*;
import java.util.*;

class pii implements Comparable {
	int first,second;
	public pii(){first = 0; second = 0;}
	public pii(int a, int b){first = a; second = b;}
	public boolean less(pii b){
		if(first < b.first)return true;
		if(first== b.first && second < b.second)return true;
		return false;
	}
	public int compareTo(Object o) {
		if (o == null) {
			return -1;
		}
		if (!(o instanceof pii)) {
			return -1;
		}
		pii p = (pii) o;
		if (p.first == first && p.second == second) {
			return 0;
		}
		return (less((pii) o) ? -1 : 1);
	}
}

public class tobi {
	public static final int MAX = 250;
	public static int numAS[] = new int[2];
	public static int numTS[] = new int[2];
	public static int numOp[] = new int[2];

	public static ArrayList<ArrayList<ArrayList<pii>>> adjList;
	public static ArrayList<ArrayList<ArrayList<Integer>>> revList;

	public static boolean productive[][] = new boolean[2][MAX];
	public static boolean traversed[][] = new boolean[MAX][MAX];

	public static void prod(int factory, int node) {
		if (productive[factory][node]) return;
		productive[factory][node] = true;
		for (int i = 0; i < revList.get(factory).get(node).size(); i++) prod(factory, revList.get(factory).get(node).get(i));
	}

	public static boolean same(int left, int right) {
		if (left == numTS[0] - 1 && right != numTS[1] - 1) return false;
		if (left != numTS[0] - 1 && right == numTS[1] - 1) return false;
		if (traversed[left][right]) return true;
		traversed[left][right] = true;

		int leftP = 0; int rightP = 0;
		while (true) {
			while (leftP < adjList.get(0).get(left).size() && !productive[0][adjList.get(0).get(left).get(leftP).second]) leftP++;
			while (rightP < adjList.get(1).get(right).size() && !productive[1][adjList.get(1).get(right).get(rightP).second]) rightP++;
			if (leftP >= adjList.get(0).get(left).size() && rightP < adjList.get(1).get(right).size()) return false;
			if (leftP < adjList.get(0).get(left).size() && rightP >= adjList.get(1).get(right).size()) return false;
			if (leftP >= adjList.get(0).get(left).size() && rightP >= adjList.get(1).get(right).size()) return true;
			if (adjList.get(0).get(left).get(leftP).first != adjList.get(1).get(right).get(rightP).first) return false;
			if (!same(adjList.get(0).get(left).get(leftP).second, adjList.get(1).get(right).get(rightP).second)) return false;
			leftP++; rightP++;
		}
	}
	
	public static void main(String[] args) {
		int numCases, from, to, label;
		Scanner sc = new Scanner(System.in);
		numCases = sc.nextInt();

		while (numCases --> 0) {
			adjList = new ArrayList<ArrayList<ArrayList<pii>>>();
			revList = new ArrayList<ArrayList<ArrayList<Integer>>>();
			for (int i = 0; i < 2; i++) {
				adjList.add(new ArrayList<ArrayList<pii>>());
				revList.add(new ArrayList<ArrayList<Integer>>());
				numAS[i] = sc.nextInt();
				numTS[i] = sc.nextInt();
				numOp[i] = sc.nextInt();
				for (int j = 0; j < numTS[i]; j++) {
					adjList.get(i).add(new ArrayList<pii>());
					revList.get(i).add(new ArrayList<Integer>());
					productive[i][j] = false;
				}
			}
			for (int i = 0; i < numTS[0]; i++) for (int j = 0; j < numTS[1]; j++) traversed[i][j] = false;
			for (int i = 0; i < 2; i++) {
				for (int j = 0; j < numAS[i]; j++) {
					from = sc.nextInt();
					to = sc.nextInt();
					label = sc.nextInt();
					pii next = new pii(label,to);
					adjList.get(i).get(from).add(next);
					revList.get(i).get(to).add(from);
				}
				for (int j = 0; j < numTS[i]; j++) {
					Collections.sort(adjList.get(i).get(j));
				}
				prod(i, numTS[i] - 1);
			}
			System.out.printf("%s\n", same(0,0) ? "eligible" : "not eligible");
		}
	}
}

File: 2023/UICPC Regional/Round #8/solutions/Outsourcing/solutions/tobi.cpp
// @EXPECTED_RESULTS@: CORRECT

#include <stdio.h>
#include <vector>
#include <algorithm>

using namespace std;

#define MAX 250
#define FOR(i,a,b) for (int i = (a); i < (b); i++)
#define sz(c) int((c).size())
typedef pair<int, int> pii;

int numAS[2], numTS[2], numOp[2];
vector<pii> adjList[2][MAX];
vector<int> revList[2][MAX];
bool productive[2][MAX], traversed[MAX][MAX];

void prod(int factory, int node) {
	if (productive[factory][node]) return;
	productive[factory][node] = true;
	FOR(i, 0, sz(revList[factory][node])) prod(factory, revList[factory][node][i]);
}

bool same(int left, int right) {
	if (left == numTS[0] - 1 && right != numTS[1] - 1) return false;
	if (left != numTS[0] - 1 && right == numTS[1] - 1) return false;
	if (traversed[left][right]) return true;
	traversed[left][right] = true;

	int leftP = 0; int rightP = 0;
	while (true) {
		while (leftP < sz(adjList[0][left]) && !productive[0][adjList[0][left][leftP].second]) leftP++;
		while (rightP < sz(adjList[1][right]) && !productive[1][adjList[1][right][rightP].second]) rightP++;
		if (leftP >= sz(adjList[0][left]) && rightP < sz(adjList[1][right])) return false;
		if (leftP < sz(adjList[0][left]) && rightP >= sz(adjList[1][right])) return false;
		if (leftP >= sz(adjList[0][left]) && rightP >= sz(adjList[1][right])) return true;
		if (adjList[0][left][leftP].first != adjList[1][right][rightP].first) return false;
		if (!same(adjList[0][left][leftP].second, adjList[1][right][rightP].second)) return false;
		leftP++; rightP++;
	}
}

int main(int argc, char **argv) {
	int numCases, from, to, label;
	scanf("%d", &numCases);

	while (numCases --> 0) {
		FOR(i, 0, 2) {
			scanf("%d%d%d", numAS+i, numTS+i, numOp+i);
			FOR(j, 0, numTS[i]) {
				adjList[i][j].clear();
				revList[i][j].clear();
				productive[i][j] = false;
			}
		}
		FOR(i, 0, numTS[0]) FOR(j, 0, numTS[1]) traversed[i][j] = false;
		FOR(i, 0, 2) {
			FOR(j, 0, numAS[i]) {
				scanf("%d%d%d", &from, &to, &label);
				pii next = pii(label,to);
				adjList[i][from].push_back(next);
				revList[i][to].push_back(from);
			}
			FOR(j, 0, numTS[i]) sort(adjList[i][j].begin(), adjList[i][j].end());
			prod(i, numTS[i] - 1);
		}
		printf("%s\n", same(0,0) ? "eligible" : "not eligible");
	}

	return 0;
}
